app/src/main/java/io/appium/settings/RecorderService.java

```diff
@@ -0,0 +1,177 @@
+import static io.appium.settings.RecorderConstant.ACTION_RECORDING_START;
+import static io.appium.settings.RecorderConstant.ACTION_RECORDING_STOP;
+
+public class RecorderService extends Service {
+    private static final String TAG = "RecorderService";
+
+    private static RecorderThread recorderThread;
+
+    public RecorderService() {
+        super();
+    }
+
+    @Override
+    public void onDestroy() {
+        Log.v(TAG, "onDestroy:");
+        if (recorderThread != null && recorderThread.isRecordingRunning()) {
+            recorderThread.stopRecording();
+        }
+        super.onDestroy();
+    }
+
+    @Nullable
+    @Override
+    public IBinder onBind(final Intent intent) {
+        return null;
+    }
+
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    @Override
+    public int onStartCommand(final Intent intent, final int flags, final int startId) {
+        Log.v(TAG, "onStartCommand:intent=" + intent);
+
+        int result = START_STICKY;
+        final String action = intent != null ? intent.getAction() : null;
+        Log.v(TAG, "onStartCommand:action string: " + action);
+        if (ACTION_RECORDING_START.equals(action)) {
+            showNotification(); // TODO is this really necessary
+
+            MediaProjectionManager mMediaProjectionManager =
+                    (MediaProjectionManager) getSystemService(Context.MEDIA_PROJECTION_SERVICE);
+
+            if (mMediaProjectionManager != null) {
+                startScreenRecord(mMediaProjectionManager, intent);
+            } else {
+                Log.e(TAG, "mMediaProjectionManager null");
+                result = START_NOT_STICKY;
+            }
+        } else if (ACTION_RECORDING_STOP.equals(action)) {
+            Log.v(TAG, "onStartCommand:intent=" + "stop");
+            stopScreenRecord();
+            result = START_NOT_STICKY;
+        } else {
+            Log.w(TAG, "onStartCommand: Unknown ACTION name");
+            result = START_NOT_STICKY;
+        }
+
+        return result;
+    }
+
+    /**
+     * start screen recording
+     */
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void startScreenRecord(MediaProjectionManager mediaProjectionManager,
+                                   final Intent intent) {
+        if (recorderThread != null) {
+            Log.v(TAG, "recorder thread is not null");
+            if (recorderThread.isRecordingRunning()) {
+                Log.v(TAG, "recording is already continuing");
+            } else {
+                Log.v(TAG, "recording is stopped");
+            }
+            return;
+        }
+
+        int resultCode = intent.getIntExtra(ACTION_RECORDING_RESULT_CODE, 0);
+        // get MediaProjection
+        final MediaProjection projection = mediaProjectionManager.getMediaProjection(resultCode,
+                intent);
+        if (projection == null) {
+            Log.e(TAG, "MediaProjection is null");
```

Comment:
such log messages are not very helpful. I usually try to keep them more informative, for example:
Cannot start screen recording: unable to retrieve MediaProjection instance

Reply:
done https://github.com/sirmordred/io.appium.settings/commit/cd56a2634a23e37076ce2f5d5ac2c005d1f52e0e